### PR TITLE
Fix match function call slowness at scale

### LINKED ISSUE #62283: [BUG] match function calls are slow at scale
**Description**

While troubleshooting a workflow that included a lot of `salt.match.compound()` calls in Jinja, it was noticed that the function call contributed quite a bit of time to the overall workflow duration. Upon further investigation, it was found that every function call in the `match` module is invoking the Salt Loader _on every run_. To make matters worse, the compound matcher is invoking the Salt Loader _again_ to pull in each distinct match type... so a call to compound matching invokes the Salt Loader **twice** on **every** call.

**Setup**

This simple script was used to test the compound match function calls at scale.

```py
import argparse
import timeit

import salt.config
import salt.matchers.compound_match


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("-n", type=int, default=1000)
    args = parser.parse_args()

    __opts__ = salt.config.minion_config("/etc/salt/minion")

    salt.matchers.compound_match.__opts__ = __opts__

    print(
        timeit.timeit(
            lambda: salt.matchers.compound_match.match("G@foo:bar"),
            number=args.n,
        )
    )


if __name__ == "__main__":
    main()
```

**Steps to Reproduce the behavior**

Running the script above results in fairly slow operation at scale.

```bash
$ python match_test.py -n 100000
67.60197884199442
```

**Expected behavior**

Changing to direct imports of the matchers yields much faster results.

```bash
$ python match_test.py -n 100000
6.912001780990977
```

<details><summary>Direct import diff for reference.</summary>

```diff
diff --git a/salt/matchers/compound_match.py b/salt/matchers/compound_match.py
index 40ef2b03fa..bf82f3a531 100644
--- a/salt/matchers/compound_match.py
+++ b/salt/matchers/compound_match.py
@@ -7,6 +7,19 @@ import logging
 import salt.loader
 import salt.utils.minions
 
+# Instead of letting this hit the lazy loader, just statically import them.
+# This is _significantly_ faster, a good 10x in a simple benchmark.
+from salt.matchers import (
+    glob_match,
+    grain_match,
+    grain_pcre_match,
+    ipcidr_match,
+    list_match,
+    pcre_match,
+    pillar_match,
+    pillar_pcre_match,
+)
+
 HAS_RANGE = False
 try:
     import seco.range  # pylint: disable=unused-import
@@ -25,7 +38,6 @@ def match(tgt, opts=None, minion_id=None):
     if not opts:
         opts = __opts__
     nodegroups = opts.get("nodegroups", {})
-    matchers = salt.loader.matchers(opts)
     if not minion_id:
         minion_id = opts.get("id")
 
@@ -34,14 +46,14 @@ def match(tgt, opts=None, minion_id=None):
         return False
     log.debug("compound_match: %s ? %s", minion_id, tgt)
     ref = {
-        "G": "grain",
-        "P": "grain_pcre",
-        "I": "pillar",
-        "J": "pillar_pcre",
-        "L": "list",
+        "G": grain_match.match,
+        "P": grain_pcre_match.match,
+        "I": pillar_match.match,
+        "J": pillar_pcre_match.match,
+        "L": list_match.match,
         "N": None,  # Nodegroups should already be expanded
-        "S": "ipcidr",
-        "E": "pcre",
+        "S": ipcidr_match.match,
+        "E": pcre_match.match,
     }
     if HAS_RANGE:
         ref["R"] = "range"
@@ -101,17 +113,11 @@ def match(tgt, opts=None, minion_id=None):
             if target_info["delimiter"]:
                 engine_kwargs["delimiter"] = target_info["delimiter"]
 
-            results.append(
-                str(
-                    matchers["{}_match.match".format(engine)](
-                        *engine_args, **engine_kwargs
-                    )
-                )
-            )
+            results.append(str(engine(*engine_args, **engine_kwargs)))
 
         else:
             # The match is not explicitly defined, evaluate it as a glob
-            results.append(str(matchers["glob_match.match"](word, opts, minion_id)))
+            results.append(str(glob_match.match(word, opts, minion_id)))
 
     results = " ".join(results)
     log.debug('compound_match %s ? "%s" => "%s"', minion_id, tgt, results)
diff --git a/salt/modules/match.py b/salt/modules/match.py
index 878ab35436..329858881e 100644
--- a/salt/modules/match.py
+++ b/salt/modules/match.py
@@ -8,9 +8,21 @@ import logging
 import sys
 from collections.abc import Mapping
 
-import salt.loader
+import salt.utils.dictupdate
 from salt.defaults import DEFAULT_TARGET_DELIM
 from salt.exceptions import SaltException
+from salt.matchers import (
+    compound_match,
+    data_match,
+    glob_match,
+    grain_match,
+    grain_pcre_match,
+    ipcidr_match,
+    list_match,
+    pcre_match,
+    pillar_match,
+    pillar_pcre_match,
+)
 
 __func_alias__ = {"list_": "list"}
 
@@ -35,9 +47,8 @@ def compound(tgt, minion_id=None):
     if minion_id is not None:
         if not isinstance(minion_id, str):
             minion_id = str(minion_id)
-    matchers = salt.loader.matchers(__opts__)
     try:
-        ret = matchers["compound_match.match"](tgt, opts=__opts__, minion_id=minion_id)
+        ret = compound_match.match(tgt, opts=__opts__, minion_id=minion_id)
     except Exception as exc:  # pylint: disable=broad-except
         log.exception(exc)
         ret = False
@@ -65,9 +76,8 @@ def ipcidr(tgt):
          - nodeclass: internal
 
     """
-    matchers = salt.loader.matchers(__opts__)
     try:
-        return matchers["ipcidr_match.match"](tgt, opts=__opts__)
+        return ipcidr_match.match(tgt, opts=__opts__)
     except Exception as exc:  # pylint: disable=broad-except
         log.exception(exc)
         return False
@@ -96,11 +106,8 @@ def pillar_pcre(tgt, delimiter=DEFAULT_TARGET_DELIM):
         .. versionadded:: 0.16.4
         .. deprecated:: 2015.8.0
     """
-    matchers = salt.loader.matchers(__opts__)
     try:
-        return matchers["pillar_pcre_match.match"](
-            tgt, delimiter=delimiter, opts=__opts__
-        )
+        return pillar_pcre_match.match(tgt, delimiter=delimiter, opts=__opts__)
     except Exception as exc:  # pylint: disable=broad-except
         log.exception(exc)
         return False
@@ -129,9 +136,8 @@ def pillar(tgt, delimiter=DEFAULT_TARGET_DELIM):
         .. versionadded:: 0.16.4
         .. deprecated:: 2015.8.0
     """
-    matchers = salt.loader.matchers(__opts__)
     try:
-        return matchers["pillar_match.match"](tgt, delimiter=delimiter, opts=__opts__)
+        return pillar_match.match(tgt, delimiter=delimiter, opts=__opts__)
     except Exception as exc:  # pylint: disable=broad-except
         log.exception(exc)
         return False
@@ -147,9 +153,8 @@ def data(tgt):
 
         salt '*' match.data 'spam:eggs'
     """
-    matchers = salt.loader.matchers(__opts__)
     try:
-        return matchers["data_match.match"](tgt, opts=__opts__)
+        return data_match.match(tgt, opts=__opts__)
     except Exception as exc:  # pylint: disable=broad-except
         log.exception(exc)
         return False
@@ -178,11 +183,8 @@ def grain_pcre(tgt, delimiter=DEFAULT_TARGET_DELIM):
         .. versionadded:: 0.16.4
         .. deprecated:: 2015.8.0
     """
-    matchers = salt.loader.matchers(__opts__)
     try:
-        return matchers["grain_pcre_match.match"](
-            tgt, delimiter=delimiter, opts=__opts__
-        )
+        return grain_pcre_match.match(tgt, delimiter=delimiter, opts=__opts__)
     except Exception as exc:  # pylint: disable=broad-except
         log.exception(exc)
         return False
@@ -211,9 +213,8 @@ def grain(tgt, delimiter=DEFAULT_TARGET_DELIM):
         .. versionadded:: 0.16.4
         .. deprecated:: 2015.8.0
     """
-    matchers = salt.loader.matchers(__opts__)
     try:
-        return matchers["grain_match.match"](tgt, delimiter=delimiter, opts=__opts__)
+        return grain_match.match(tgt, delimiter=delimiter, opts=__opts__)
     except Exception as exc:  # pylint: disable=broad-except
         log.exception(exc)
         return False
@@ -237,9 +238,8 @@ def list_(tgt, minion_id=None):
     if minion_id is not None:
         if not isinstance(minion_id, str):
             minion_id = str(minion_id)
-    matchers = salt.loader.matchers(__opts__)
     try:
-        return matchers["list_match.match"](tgt, opts=__opts__, minion_id=minion_id)
+        return list_match.match(tgt, opts=__opts__, minion_id=minion_id)
     except Exception as exc:  # pylint: disable=broad-except
         log.exception(exc)
         return False
@@ -263,9 +263,8 @@ def pcre(tgt, minion_id=None):
     if minion_id is not None:
         if not isinstance(minion_id, str):
             minion_id = str(minion_id)
-    matchers = salt.loader.matchers(__opts__)
     try:
-        return matchers["pcre_match.match"](tgt, opts=__opts__, minion_id=minion_id)
+        return pcre_match.match(tgt, opts=__opts__, minion_id=minion_id)
     except Exception as exc:  # pylint: disable=broad-except
         log.exception(exc)
         return False
@@ -289,10 +288,9 @@ def glob(tgt, minion_id=None):
     if minion_id is not None:
         if not isinstance(minion_id, str):
             minion_id = str(minion_id)
-    matchers = salt.loader.matchers(__opts__)
 
     try:
-        return matchers["glob_match.match"](tgt, opts=__opts__, minion_id=minion_id)
+        return glob_match.match(tgt, opts=__opts__, minion_id=minion_id)
     except Exception as exc:  # pylint: disable=broad-except
         log.exception(exc)
         return False
```
</details>

As far as I can tell, there's no reason to invoke the Salt Loader for known-present matchers since we're not relying on `__virtual__()` to selectively load matchers or storing the loaded _matchers_ in a location (like `__salt__`) where they can be invoked without incurring loader overhead each time.

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```yaml
Salt Version:
               Salt: 3004.2
 
Dependency Versions:
               cffi: Not Installed
           cherrypy: Not Installed
           dateutil: 2.7.3
          docker-py: Not Installed
              gitdb: 2.0.5
          gitpython: 2.1.11
             Jinja2: 2.10
            libgit2: Not Installed
           M2Crypto: Not Installed
               Mako: Not Installed
            msgpack: 0.5.6
       msgpack-pure: Not Installed
       mysql-python: Not Installed
          pycparser: Not Installed
           pycrypto: 2.6.1
       pycryptodome: 3.6.1
             pygit2: Not Installed
             Python: 3.7.3 (default, Jan 22 2021, 20:04:44)
       python-gnupg: Not Installed
             PyYAML: 3.13
              PyZMQ: 17.1.2
              smmap: 2.0.5
            timelib: Not Installed
            Tornado: 4.5.3
                ZMQ: 4.3.1
 
Salt Extensions:
 saltext.prometheus: 1.1.1
 
System Versions:
               dist: debian 10 buster
             locale: UTF-8
            machine: x86_64
            release: 4.19.0-20-amd64
             system: Linux
            version: Debian GNU/Linux 10 buster
```
</details>

**Additional context**
The slowdown here isn't significant enough to notice with "normal" targeting. The match functions would need to be called many times in quick succession, so this really affects function calls in Jinja and custom code.

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
This PR changes match functions from loader access to direct imports of the matchers in order to yield much faster results.

### What issues does this PR fix or reference?
Fixes: #62283 

### Previous Behavior
See issue for details.

### New Behavior
See issue for details.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Review by whytewolf — DISMISSED** [review]

**Comment by nicholasmhughes** [comment]
Question for the class...

Why would `log.debug()` be a (somewhat relatively) expensive operation when not in `DEBUG` and using string interpolation with standard types?

Removing some debug lines in this branch further reduces run time from:

```
$ python match_test.py -n 100000
6.784708013990894
```

to:

```
$ python match_test.py -n 100000
1.6946497799362987
```

<details><summary>Diff for reference</summary>

```diff
diff --git a/salt/matchers/compound_match.py b/salt/matchers/compound_match.py
index f2be8a5b57..a4068932d3 100644
--- a/salt/matchers/compound_match.py
+++ b/salt/matchers/compound_match.py
@@ -43,7 +43,7 @@ def match(tgt, opts=None, minion_id=None):
     if not isinstance(tgt, str) and not isinstance(tgt, (list, tuple)):
         log.error("Compound target received that is neither string, list nor tuple")
         return False
-    log.debug("compound_match: %s ? %s", minion_id, tgt)
+    #log.debug("compound_match: %s ? %s", minion_id, tgt)
     ref = {
         "G": grain_match.match,
         "P": grain_pcre_match.match,
@@ -119,7 +119,7 @@ def match(tgt, opts=None, minion_id=None):
             results.append(str(glob_match.match(word, opts, minion_id)))
 
     results = " ".join(results)
-    log.debug('compound_match %s ? "%s" => "%s"', minion_id, tgt, results)
+    #log.debug('compound_match %s ? "%s" => "%s"', minion_id, tgt, results)
     try:
         return eval(results)  # pylint: disable=W0123
     except Exception:  # pylint: disable=broad-except
diff --git a/salt/matchers/grain_match.py b/salt/matchers/grain_match.py
index 0dd541b3e2..f71334183d 100644
--- a/salt/matchers/grain_match.py
+++ b/salt/matchers/grain_match.py
@@ -17,7 +17,7 @@ def match(tgt, delimiter=DEFAULT_TARGET_DELIM, opts=None, minion_id=None):
     if not opts:
         opts = __opts__
 
-    log.debug("grains target: %s", tgt)
+    #log.debug("grains target: %s", tgt)
     if delimiter not in tgt:
         log.error("Got insufficient arguments for grains match statement from master")
         return False
diff --git a/salt/utils/data.py b/salt/utils/data.py
index 0b3634d842..002199527d 100644
--- a/salt/utils/data.py
+++ b/salt/utils/data.py
@@ -972,12 +972,12 @@ def subdict_match(
         else:
             matchstr = delimiter.join(splits[idx:])
             match = traverse_dict_and_list(data, key, {}, delimiter=delimiter)
-        log.debug(
-            "Attempting to match '%s' in '%s' using delimiter '%s'",
-            matchstr,
-            key,
-            delimiter,
-        )
+        #log.debug(
+        #    "Attempting to match '%s' in '%s' using delimiter '%s'",
+        #    matchstr,
+        #    key,
+        #    delimiter,
+        #)
         if match == {}:
             continue
         if isinstance(match, dict):
```
</details>

File: salt/matchers/compound_match.py [hunk 1]
```diff
@@ -25,7 +37,6 @@ def match(tgt, opts=None, minion_id=None):
     if not opts:
         opts = __opts__
     nodegroups = opts.get("nodegroups", {})
-    matchers = salt.loader.matchers(opts)
```
Comment (OrangeDog) [inline salt/matchers/compound_match.py:28]: Did passing `opts` do anything that is now not accounted for?

Comment (nicholasmhughes) [reply salt/matchers/compound_match.py:28]: We're already passing `opts` directly into the function, so I think we're covered.

**Comment by waynew** [comment]
> Question for the class...
> 
> Why would `log.debug()` be a (somewhat relatively) expensive operation when not in `DEBUG` and using string interpolation with standard types?

Without digging into the code... my recollection is that in the Python logging module that it should *not* do string interpolation if the logging is not enabled for that level. In theory it should bail out. You could also check the time with log calls that *aren't* using any interpolation. But I'm pretty sure that what's happening is just that it's incurring the cost of the function calls, which is non-zero (and in fact, one of the most expensive things to do in Python).

**Comment by nicholasmhughes** [comment]
I dug into this a little further last week and benchmarked the salt logging class against the default in a super simple test case.

<details><summary>nick_py_logger.py</summary>

```py
import argparse
import logging
import timeit

log = logging.getLogger(__name__)


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("-n", type=int, default=1000)
    args = parser.parse_args()

    one = "monkey"
    two = ["donkey"]

    print(
        timeit.timeit(
            lambda: log.debug("stuff: %s ? %s", one, two),
            number=args.n,
        )
    )


if __name__ == "__main__":
    main()
```
</details>

```bash
$ python nick_py_logger.py -n 100000
0.01669288403354585
```

<details><summary>nick_salt_logger.py</summary>

```py
import argparse
import logging
import timeit

from salt._logging.impl import (SaltLoggingClass, PROFILE, TRACE, QUIET, GARBAGE, setup_temp_handler, get_temp_handler)


if logging.getLoggerClass() is not SaltLoggingClass:

    # Import pip._internal which itself will install it's own custom logging handler
    # we want to override that handler with ours
    try:
        import pip._internal.utils._log as pip_log_module  # pylint: disable=no-name-in-module,import-error
    except ImportError:
        pip_log_module = None

    logging.setLoggerClass(SaltLoggingClass)
    logging.addLevelName(QUIET, "QUIET")
    logging.addLevelName(PROFILE, "PROFILE")
    logging.addLevelName(TRACE, "TRACE")
    logging.addLevelName(GARBAGE, "GARBAGE")
    if pip_log_module is not None:
        # Let's make newer versions of pip work by patching SaltLoggingClass to
        # add a verbose method which is what pip expects
        SaltLoggingClass.verbose = SaltLoggingClass.debug

    if not logging.root.handlers:
        # No configuration to the logging system has been done so far.
        # Set the root logger at the lowest level possible
        logging.root.setLevel(GARBAGE)

        # Add a permanent null handler so that we never get messages like:
        #   No handlers could be found for logger 'foo'
        setup_temp_handler()
        logging.root.addHandler(get_temp_handler())


# Now that we defined the default logging logger class, we can instantiate our logger
# DO NOT MOVE THIS
log = logging.getLogger(__name__)


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("-n", type=int, default=1000)
    args = parser.parse_args()

    one = "monkey"
    two = ["donkey"]

    print(
        timeit.timeit(
            lambda: log.debug("stuff: %s ? %s", one, two),
            number=args.n,
        )
    )


if __name__ == "__main__":
    main()
```
</details>

```bash
$ python nick_salt_logger.py -n 100000
0.6871055489173159
```

40x slower just seems excessive for a wrapper that adds some basic info and then offloads to the default logging class method to do the bulk of the work.

**Comment by OrangeDog** [comment]
In general, Python is slow. Repeating the level check before adding "some basic info" may speed it up a bit, but it's hard to predict.

**Comment by Ch3LL** [comment]
ping @cro any chance you can review here. I know you initially added the feature to allow matchers to be loaded by Salt's loader.

**Comment by cro** [comment]
This will prevent the loading of any custom matchers completely, only the matchers hardcoded will be permitted.  If any Salt users depend on custom matchers (like overloading the glob matcher with different semantics that fit their environment better, and letting the loader system distribute the module), this PR will break their setup.

I'm not specifically against the idea of importing them directly for a performance boost, is there a way to do this while keeping the ability to load custom matchers?

**Comment by nicholasmhughes** [comment]
The only way I can think of to solve that problem is to create `__matchers__` as global dunder and handle it like we do with `__salt__`. That way, it'd only be loaded up once instead of during each call to the function. Thoughts on this approach?

I mocked up something this morning to test out the idea, and it seems to work. However, pytest-salt-factories will probably have to recognize `__matchers__` as a legitimate dunder in order to ease testing of the functions with this new setup. @s0undt3ch , can you confirm?

**Comment by nicholasmhughes** [comment]
:point_up_2: @cro @garethgreenaway @s0undt3ch

**Comment by cro** [comment]
> have to recognize __matchers__ as a legitimate dunder in order to ease testing of the functions with this new setup. 
> @s0undt3ch , can you confirm?

This will be helpful, and I thought it had already been done, but checking the code reveals it's not there.  It's not hard to do, but it is a fair amount of code to modify.

**Comment by s0undt3ch** [comment]
> The only way I can think of to solve that problem is to create `__matchers__` as global dunder and handle it like we do with `__salt__`. That way, it'd only be loaded up once instead of during each call to the function. Thoughts on this approach?
> 
> I mocked up something this morning to test out the idea, and it seems to work. However, pytest-salt-factories will probably have to recognize `__matchers__` as a legitimate dunder in order to ease testing of the functions with this new setup. @s0undt3ch , can you confirm?

Yes, we would need to update salt-factories.

**Comment by s0undt3ch** [comment]
I'd prefer not to create another dunder global if we can avoid it.

**Comment by Ch3LL** [comment]
@s0undt3ch can you elaborate why you would not want to do this? This would ensure we are not invoking the loader on each matcher call, which would mean more speed.

**Comment by nicholasmhughes** [comment]
@s0undt3ch , do you have an alternative? :point_up:

**Comment by nicholasmhughes** [comment]
All, I moved to storing the matchers in `__context__`, so they should only be loaded once. It's not quite as fast... but should allow for the matcher overrides to continue as before. Please let me know if this method is acceptable.

**Comment by nicholasmhughes** [comment]
re-run pr-freebsd-131-amd64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-centos-7-x86_64-py3-tcp-pytest

**Comment by cro** [comment]
Thank you @nicholasmhughes !  I'm fine with that, good idea.

**Review by whytewolf — DISMISSED** [review]

**Comment by Ch3LL** [comment]
There is a recent merge conflict that needs to be resolved.

**Comment by nicholasmhughes** [comment]
Fixed

**Comment by nicholasmhughes** [comment]
re-run pr-ubuntu-2004-amd64-py3-pycryptodome-pytest

**Review by garethgreenaway — APPROVED** [review]

**Comment by nicholasmhughes** [comment]
re-run re-run pr-ubuntu-2004-amd64-py3-pycryptodome-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-ubuntu-2004-amd64-py3-pycryptodome-pytest

**Review by dmurphy18 — APPROVED** [review]

**Review by s0undt3ch — APPROVED** [review]

**Comment by cmcmarrow** [comment]
https://github.com/saltstack/salt/pull/64048/files

**Pull request merged**